PSR-16 compatible cache interface

burzum opened the pull request.

burzum committed:
Started to work on the PSR6 / PSR16 implementation

burzum committed:
Made cache tests run without fatal errors again

burzum committed:
Adding writeMany() back

burzum committed:
Updating the RedisEngine

burzum committed:
Refactoring more cache engines for PSR6

burzum committed:
Fixing the Cache tests

stickler-ci committed:
Fixing style errors.

burzum committed:
Adding CacheException and working on the Cache API

burzum committed:
Merge branch '3.next-cache' of github.com:burzum/cakephp into 3.next-cache

# Conflicts:
#	src/Cache/Cache.php
#	src/Cache/CacheEngine.php

stickler-ci committed:
Fixing style errors.

burzum committed:
Fixing apcu and wincache engines get()

burzum committed:
Fixing Cache::read()

Needs to call the deprecated read() method on the engines, because they take care of returning this in a BC fashion

burzum committed:
Minor doc block changes

dereuromark commented on src/Cache/Cache.php:
```diff
@@ -388,7 +442,37 @@ public static function readMany($keys, $config = 'default')
+     * @return array An array containing, for each of the given $keys, the cached data or false if cached data could not be
+     * retrieved.
+     */
+    public function getMultiple($keys, $default = null, $config = 'default')
```
should we have array typehint here?

ADmad replied:
https://github.com/php-fig/simple-cache/blob/master/src/CacheInterface.php#L58
So the typehint in docblock should be changed from `array` to `iterable`. Similarly for all `*Multiple()` methods.

dereuromark commented on src/Cache/Cache.php:
```diff
@@ -486,6 +571,35 @@ public static function deleteMany($keys, $config = 'default')
+     * @return bool
+     * couldn't be removed
+     */
+    public static function deleteMultiple($keys, $config = 'default')
```
should we have array typehint here?

dereuromark commented on src/Cache/Cache.php:
```diff
@@ -486,6 +571,35 @@ public static function deleteMany($keys, $config = 'default')
+     * @param array $keys Array of cache keys to be deleted
+     * @param string $config name of the configuration to use. Defaults to 'default'
+     * @return bool
+     * couldn't be removed
```
What does this comment mean? Should this be above the params instead with a more clear message?

dereuromark commented on src/Cache/CacheEngine.php:
```diff
@@ -295,4 +211,165 @@ protected function warning($message)
+     *   MUST be thrown if $values is neither an array nor a Traversable,
+     *   or if any of the $values are not a legal value.
+     */
+    public function setMultiple($values, $ttl = null)
```
You use iterable here on doc block, but typehint is not?

markstory commented:
I'll try to fix a few of the failing tests in the next day or so.

burzum committed:
Adding the methods back and removing the BC trait

Reason for this is it turned out that there are some to specific cases to handle it with the trait alone. :(

stickler-ci committed:
Fixing style errors.

markstory committed:
Reduce duplication in FileEngine and fix ttl issue.

FileEngine did not respect the ttl parameter. I've added tests for this
scenario.

markstory commented on src/Cache/Engine/RedisEngine.php:
```diff
@@ -123,47 +121,84 @@ protected function _connect()
+            $data = serialize($data);
+        }
+
+        $duration = $this->_config['duration'];
```
We aren't using the ttl parameter here.

burzum replied:
Fixed.

markstory commented on src/Cache/Engine/XcacheEngine.php:
```diff
@@ -71,11 +71,63 @@ public function init(array $config = [])
+     *                                     for it or let the driver take care of that.
+     * @return bool True if the data was successfully cached, false on failure
+     */
+    public function set($key, $value, $ttl = null)
```
This method isn't using the ttl parameter.

burzum replied:
Fixed it, hopefully correctly.

markstory committed:
Fix default value handling in FileEngine::get()

markstory commented on src/Cache/Engine/NullEngine.php:
```diff
@@ -113,4 +122,20 @@ public function clearGroup($group)
+     */
+    public function get($key, $default = null)
+    {
+        return false;
```
Shouldn't this return `$default`?

burzum replied:
Hm... whats the purpose of the Null engine? In which scenario do we want the default instead of false?

markstory commented on src/Cache/Engine/RedisEngine.php:
```diff
@@ -123,47 +121,84 @@ protected function _connect()
+            return $result === false ? null : $result;
+        }
+
+        return $value === false ? null : $value;
```
We aren't using the default value in this method, we should return that on failures.

burzum committed:
TTL for XcacheEngine

burzum committed:
Returning default values for NullEngine and RedisEngine

burzum committed:
Fixing cache tests

stickler-ci committed:
Fixing style errors.

markstory committed:
Add missing coverage to fileengine.

markstory commented on src/Cache/Cache.php:
```diff
@@ -500,6 +632,19 @@ public static function clear($check = false, $config = 'default')
+     * @param string $config name of the configuration to use. Defaults to 'default'
+     * @return bool
+     */
+    public function clearExpired($config = 'default')
```
If this isn't part of the PSR interfaces why are we adding it? It won't do anything on most engines as they auto-expire data.

markstory committed:
Redis extension is working again in php7.2

markstory commented on src/Cache/CacheEngine.php:
```diff
@@ -171,31 +110,6 @@ public function readMany($keys)
-     * @param bool $check if true will check expiration, otherwise delete all
-     * @return bool True if the cache was successfully cleared, false otherwise
-     */
-    abstract public function clear($check);
```
I'm not sure we can change the `clear()` signature without breaking compatibility as it could have been implemented by userland engines.

markstory replied:
I was thinking some mote about this last night. We could add the new methods required by the PSR interfaces now, but not implement the actual interfaces until 4.0 when we can make the breaking changes to clear().

ADmad replied:
@markstory In that case wouldn't it be better to also not change the return value for engines in case of cache miss from current `false` to `null`? It could potentially cause issues with current applications doing strict checks.

markstory replied:
I don't think we should change return values for the existing write/read methods as it could cause compatibility issues. However, we sould implement get(), set(), and has() in a way that will be compatible with the PSR interfaces.

markstory commented:
Going to close this as the wrapper approach implemented in #12520 is lower risk for 3.x.
